source/mir/random/engine/mersenne_twister.d

```diff
@@ -90,6 +106,198 @@ struct MersenneTwisterEngine(UIntType, size_t w, size_t n, size_t m, size_t r,
 
     /++
+    Constructs a MersenneTwisterEngine object.
+
+    Note that  `MersenneTwisterEngine([123])` will not result in
+    the same initial state as `MersenneTwisterEngine(123)`.
+    +/
+    this(scope Slice!(Contiguous, [1], const(UIntType)*) slice) @safe pure nothrow @nogc
+    {
+        static if (is(UIntType == uint))
+        {
+            enum UIntType f2 = 1664525u;
+            enum UIntType f3 = 1566083941u;
+        }
+        else static if (is(UIntType == ulong))
+        {
+            enum UIntType f2 = 3935559000370003845uL;
+            enum UIntType f3 = 2862933555777941757uL;
+        }
+        else
+            static assert(0, "init by slice only supported if UIntType is uint or ulong!");
+
+        data[$-1] = cast(UIntType) (19650218u & max);
+        foreach_reverse (size_t i, ref e; data[0 .. $-1])
+        {
+            e = f * (data[i + 1] ^ (data[i + 1] >> (w - 2))) + cast(UIntType)(n - (i + 1));
+            static if (max != UIntType.max)
+                e &= max;
+        }
+        index = n-1;
+        if (slice.length == 0)
+        {
+            opCall();
+            return;
+        }
+
+        size_t i = n - 2;
+        size_t j = 0;
+        immutable key_length = slice.length;
+        const(UIntType)[] field = slice.field;
+        foreach_reverse(_; 0 .. (n > key_length ? n : key_length))
+        {
+            data[i] = (data[i] ^ ((data[i+1] ^ (data[i+1] >> (w - 2))) * f2))
+                + cast(UIntType)(field[j] + j); /* non linear */
+            static if (max != UIntType.max)
+                data[i] &= max;
+            if (--i > n)
+            {
+                data[$ - 1] = data[0];
+                i = n - 2;
+            }
+            ++j;
+            if (j >= key_length)
+                j = 0;
+        }
+
+        foreach_reverse(_; 0 .. n-1)
+        {
+            data[i] = (data[i] ^ ((data[i+1] ^ (data[i+1] >> (w - 2))) * f3))
+                - cast(UIntType)(n - (i + 1)); /* non linear */
+            static if (max != UIntType.max)
+                data[i] &= max;
+            if (--i > n)
+            {
+                data[$ - 1] = data[0];
+                i = n - 2;
+            }
+        }
+        data[$-1] = (cast(UIntType)1) << ((UIntType.sizeof * 8) - 1); /* MSB is 1; assuring non-zero initial array */
+        opCall();
+    }
+
+    /// ditto
+    this()(scope const(UIntType)[] array) @safe pure nothrow @nogc
+    {
+        import mir.ndslice.slice: sliced;
+        this(array.sliced);
+    }
+
+    /++
+    Constructs a MersenneTwisterEngine object.
+
+    This method copies the values from the range into a temporary buffer.
+    If you can pass an array or slice directly that is more efficient.
+
+    A range that is detected as infinite by `std.range.primitives.isInfinite`
+    will be treated as though its length were `n`.
+    +/
+    this(Range)(scope Range range)
+    if (isInputRange!Range && is(Unqual!(ElementType!Range) == UIntType))
```

Comment:
this should be three line of code function:
```
import std.algorithm.mutation: copy;
import std.range: take;
this(buffer[0 .. (range.take(n).copy(buffer[]).length].sliced);
```

Reply:
Wouldn't that introduce a dependency on druntime?

Reply:
it would not